### PR TITLE
Bug 1239072 - Authentication Settings + Settings Refactor

### PR DESCRIPTION
**Pull request by sleroux**
This PR refactors the SettingsTableViewController into a reusable component that can be used for the TouchID/Passcode settings screen. The work includes:

* Extracting the class definitions for App-wide settings into AppSettingsOptions.swift
* Creating a new AppSettingsTableViewController subclass of the existing SettingsTableViewController class to extract specific logic.
* New AuthenticationSettingsViewController that also leverages the simplified SettingsTableViewController along with specific option class declarations.
* Addition of an 'enabled flag on a Setting which is used to enable/disable settings in the Authentication Setting screen.
* Scaffolding for preference keys for setting up passcode require intervals.
* Feature flag for the authentication settings work.

### CODE REVIEW

File: Client/Frontend/AuthenticationManager/AuthenticationSettingsViewController.swift [hunk 1]
```diff
@@ -0,0 +1,166 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (thebnich) [inline Client/Frontend/AuthenticationManager/AuthenticationSettingsViewController.swift]: Since this is just another Settings view controller, and since there are no other classes in the `AuthenticationManager` group you're creating, is there any reason you don't want to just put this in Client > Frontend > Settings with the others?

**Comment by thebnich** [comment]
I'm seeing strange behavior when I try this PR locally in the simulator/iPod touch. When I click Touch ID & Passcode, clicking any of the 4 rows just highlights the row without doing anything. All I can do is toggle the Use Touch ID switch, which doesn't seem to do anything.

I assume this is because the simulator/iPod touch doesn't support Touch ID. I think we'll want to add some kind of support detection so we can hide this setting on unsupported devices.

File: Client/Frontend/AuthenticationManager/AuthenticationSettingsViewController.swift (hunk 1 above)
Comment (sleroux) [inline Client/Frontend/AuthenticationManager/AuthenticationSettingsViewController.swift]: Eventually there is going to be other view controllers such as the passcode configuration and interval screens. I guess it depends on how we want to group up the classes. We can either make it revolve a theme (Authentication) or have it grouped by function (Settings, Views, ViewControllers).

**Comment by sleroux** [comment]
So all of the rows in the AuthenticationSettingsViewController actually don't do anything yet. This PR is just to setup the options and UI. I'm working on building the passcode controller next. I've also included the feature flag so we can land this without exposing broken behavior to the user.

**Comment by thebnich** [comment]
OK, I see -- I should have read more first :)

File: Client/Frontend/Settings/SettingsTableViewController.swift [hunk 2]
```diff
@@ -857,8 +247,14 @@ class SettingsTableViewController: UITableViewController {
 
     override func viewWillAppear(animated: Bool) {
         super.viewWillAppear(animated)
+
+        settings = generateSettings()
+
         NSNotificationCenter.defaultCenter().addObserver(self, selector: "SELsyncDidChangeState", name: NotificationProfileDidStartSyncing, object: nil)
         NSNotificationCenter.defaultCenter().addObserver(self, selector: "SELsyncDidChangeState", name: NotificationProfileDidFinishSyncing, object: nil)
+        NSNotificationCenter.defaultCenter().addObserver(self, selector: "SELfirefoxAccountDidChange", name: NotificationFirefoxAccountChanged, object: nil)
```
Comment (thebnich) [inline Client/Frontend/Settings/SettingsTableViewController.swift]: Should these be moved into `AppSettingsTableViewController` since these are specific to a single settings implementation?

File: Client/Frontend/Settings/SettingsTableViewController.swift [hunk 3]
```diff
@@ -947,30 +353,20 @@ class SettingsTableViewController: UITableViewController {
...
         }
-        return nil
     }
 
     override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
-        //make account/sign-in and close private tabs rows taller, as per design specs
-        if indexPath.section == 0 && indexPath.row == 0 {
+        // Make account/sign-in and close private tabs rows taller, as per design specs.
+        let section = settings[indexPath.section]
+        if let setting = section[indexPath.row] as? BoolSetting where setting.prefKey == "settings.closePrivateTabs" {
```
Comment (thebnich) [inline Client/Frontend/Settings/SettingsTableViewController.swift]: Same question here.

**Comment by thebnich** [comment]
Do you anticipate that other settings VCs (i.e., `SearchSettingsTableViewController` and `ClearPrivateDataTableViewController`) will also subclass `SettingsTableViewController` eventually? If not, what determines whether a setting VC will or will not subclass `SettingsTableViewController`?

**Comment by sleroux** [comment]
I can see `SearchSettingsTableViewController` and `ClearPrivateDataTableViewController` becoming subclasses of `SettingsTableViewController`. In general I think any 'settings' page that shares a similar look and feel as the main settings of the app would be subclasses of the `SettingsTableViewController`. For any new options that require a custom layout, its pretty easy to slot that in as another 'Setting' subclass. That way any custom work we do for a particular screen is easily transferable to other settings screen if we need it.

In other words, I see SettingsTableViewController being a part of our toolkit that we can share with the UX team when it comes to designing new settings screens. From the existing designs it looks like there is a common theme to them and creating this subclass will allow us to easily map the UX requirements to code.

File: Client/Frontend/Settings/SettingsTableViewController.swift (hunk 2 above)
Comment (sleroux) [inline Client/Frontend/Settings/SettingsTableViewController.swift]: In this case it seems like its only specific to the app settings but I feel like this has the potential to be used in many settings screens. Generally settings are tied to users profiles so I can see this being something we want. The callbacks right now only reload the table views which I think is OK but if we get more complicated logic for these callbacks down the road I'd say we break this out into an Account-based settings base class.
